tasks/simplification.py

```diff
@@ -0,0 +1,227 @@
+import os
+from luigi import Task, WrapperTask, Parameter, LocalTarget
+from util import PostgresTarget, shell
+from meta import CurrentSession
+
+OBSERVATORY_SCHEMA = 'observatory'
+DEFAULT_GEOMFIELD = 'the_geom'
+TMP_DIRECTORY = 'tmp'
+SIMPLIFICATION_DIRECTORY = 'simplification'
+SIMPLIFIED_SUFFIX = '_simpl'
+SKIPFAILURES_NO = 'no'  # Avoids https://trac.osgeo.org/gdal/ticket/6803
+SKIPFAILURES_YES = 'yes'
+DEFAULT_M_RETAIN_PERCENTAGE = '50'  # [0-100] Percentage of removable vertices retained by mapshaper
+DEFAULT_P_RETAIN_FACTOR = '50'  # Retain factor used for PostGIS simplification (this is NOT a percentage) \
+# The higher the retain factor, the lower the simplification
+UNKNOWN_RETAIN_FACTOR = '?'
+
+
+def tmp_directory(schema, table):
+    return os.path.join(TMP_DIRECTORY, SIMPLIFICATION_DIRECTORY,
+                        '{schema}.{table}'.format(schema=schema, table=table))
+
+
+def shp_filename(schema, table, suffix=''):
+    return '{schema}.{table}{suffix}.shp'.format(schema=schema, table=table, suffix=suffix)
+
+
+class ExportShapefile(Task):
+    schema = Parameter()
+    table = Parameter()
+    skipfailures = Parameter(default=SKIPFAILURES_NO)
+
+    def run(self):
+        self.output().makedirs()
```

Comment:
The Shapefile is not only one file but a bunch of files. `temporary_path` generates a path like `tmp/simplification/es.ine.RawGeometry__99914b932b/RawGeometry__99914b932b.shp-luigi-tmp-4386797297` that `ogr2ogr` interprets as a directory rather than a file. If we use the generated temp directory for the output, `ogr2ogr` will use the table name as the file name, which is lowercase (for example `rawgeometry__xxx` instead of `RawGeometry__xxx`) and it breaks all the other tasks of the simplification unless we explicitly set the input parameters to be lowercase which is ugly

Reply:
And what about just using a fixed name inside the directory? Something like `shapefile=os.path.join(temp_dir, 'simplify.shp')`. We have to modify the rest of tasks as well, to do the same.

Or take a look at `AtomicLocalFile` and see if it works. IIRC, the problem there was that the context manager (`with ...`) returned a file object rather than a path.

Just thinking because I'd like to impose a rule to use the atomic files/directories for new code. It can be more work in some circumstances, but it guarantees safety for parallel running workers.

Reply:
Also, if we are not going to be overwriting, maybe it makes sense to do a directory per-task, for consistency with the rest of the code. Right now, it seems we are sharing the directory, and putting files with different names inside.

Reply:
Of course it makes sense addressing the atomic writes problem. I have modified it to use [the recommended `temporary_path()`](http://luigi.readthedocs.io/en/stable/luigi_patterns.html#atomicwrites).

As the Shapefile is formed by several files I have implemented the solution based on the directory containing the Shapefile. The main drawback is that the `output` for the task is a `LocalTarget` pointing to the directory containing the Shapefile and `temporary_path()` creates only the parent directories (not the Shapefile directory itself), so we need to manually create the directory for the Shapefile doing `self.output().fs.mkdir(temp_path)`. Besides that, the temporary directory (with Luigi managing the renaming and deleting tasks) and the contained Shapefile behave as expected